### PR TITLE
Design doc - federated service discovery.

### PR DESCRIPTION
**Pull request by madhusudancs**
cc @nikhiljindal @ArtfulCoder @thockin @davidopp 

[Analytics]()

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/25107)

### CODE REVIEW

File: docs/design/federated-service-discovery.md [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+DNS server instance for domain name resolution. The rest of this doc refers to
+this DNS server instance as **`K8S-DNS`**.
+
+**`K8S-DNS`** runs an instance of `SkyDNS` along with its own `etcd` and
+`kube2sky` controller in a Kubernetes pod. Each of these components run in
+their own container within the pod. `kube2sky` watches the services in the
+cluster and programs the etcd backing the `SkyDNS`. In addition to these
+components, another container running an HTTP server, called stubzone server,
+is introduced into this pod to facilitate Ubernetes service discovery. This
+HTTP server exposes a single endpoint that allows programming the
```
Comment (brendandburns) [inline docs/design/federated-service-discovery.md]: How is this API going to be secured?

Comment (brendandburns) [inline docs/design/federated-service-discovery.md]: ah, I see the TLS client cert a little later.  I'd add that to the diagram.  Maybe a lock or something.

**Comment by madhusudancs** [comment]
# Federated Service Discovery using CNAME chains.

This is an alternative design proposed for Federated Service Discovery
in Ubernetes. This was all designed by @quinton-hoole, I am just
putting down all the discussions we had into a written form.

## Clarification on Terminology

The term `zone` in this document refers to `availability zones` as it
is used in the public cloud infrastructure jargon and should not be
confused with `DNS zones`. Reference to `DNS zones` in this document
are always qualified with the `DNS` prefix to help disambiguate
between the two `zones`.

# How does it work?

## `kube-dns` Changes

`kube-dns` looks at the name in the DNS query and proceeds as normal
to check if it has answers for the query. If it finds an answer, it
returns the answers to the client as it does today.

However, when it fails to find an answer, `kube-dns` is modified to
check the domain name suffix in the queried name. If the domain name
suffix matches the suffix for a federated service, `kube-dns` sends a
CNAME response with the zone and region names inserted in original
name in the query. See the [Internal Clients](## Internal Clients)
section for an example.

A new functionality that enables `kube-dns` to determine the zone and
region it is running in by reading the
`failure-domain.beta.kubernetes.io/zone` and
`failure-domain.beta.kubernetes.io/region` node labels will be built
into `kube-dns`. This allows `kube-dns` to send a CNAME response
containing the zone name.

## Cloud DNS

A Cloud DNS service such as Google Cloud DNS is programmed to respond
to the name queries containing zone and region names. If a service
shard for the requested service does not exist in the requested zone,
Cloud DNS sends a CNAME replacing the zone and region names with just
the region name. This chain  of CNAMEs are followed until the client
finds an A record for the service.

We effectively build a tree of CNAME and A records in the Cloud DNS
to handle cross-cluster service discovery.

## Internal Clients

It's easier to discuss this design with an example. Let's assume that:

  1. There is a federated service with the name `mysvc` in namespace
     `myns` in the federation `myfed`.
  2. `u7s.io` is a registered domain name and its nameservers are
     configured to point to Google Cloud DNS.
  3. There are two regions: `r1` and `r2`.
  4. There are two zones per region: `z1` and `z2` in `r1`, and `z3`
     and `z4` in `r2`.
  5. There are a total of four Kubernetes clusters: `k8sc1` running in
     `z1`, `k8sc2` running in `z2`, `k8sc3` running in `z3`, `k8sc4`
     running in `z4` that are part of the federation `myfed` with each
     cluster running in a separate zone.
  5. `mysvc` has a single shard running in `k8sc4` in `z4`.

A client running inside a pod in a Kubernetes cluster `k8sc1` running
in zone `z1` in region `r1` sends a DNS request to `kube-dns` for
`mysvc.myns.myfed.svc.u7s.io.`. `k8s-dns` fails to find an answer to
this query. It figures out that this is a query for a federated
service by looking at the domain name suffix, so it sends a CNAME
response with `mysvc.myns.myfed.svc.z1.r1.federation.` as the
canonical name. The client's resolver library then looks up this
domain name again and the request recurses out and ends up in Google
Cloud DNS which is the authoritative server for this domain.

Google Cloud DNS looks up for this name and finds out that there is a
CNAME configured for this domain name (i.e. the federated service
`mysvc`) because its shard doesn't exist in any of the clusters in
zone `z1`. It sends a CNAME pointing to
`mysvc.myns.myfed.svc.r1.u7s.io.`. The client's resolver library looks
up for this name again, which recurses out to Google Cloud DNS, which
again sends a CNAME pointing to `mysvc.myns.myfed.svc.u7s.io.`. The
client's resolver library then looks up for this new name once again
and Google Cloud DNS finds a A records for this name and sends that
out as a response.

The client then connects to the federated service `mysvc` by
connecting to one of the IP addresses returned as A records.

## External Clients.

Service discovery for external clients follows the same approach,
except that the external clients have to provide the fully qualified
domain name of the service with the registered domain name as the
suffix, e.g. `u7s.io.`.

cc @quinton-hoole @thockin @ArtfulCoder @davidopp @nikhiljindal @brendandburns

**Comment by quinton-hoole** [comment]
@madhusudancs Thanks so much for writing up my babble into such a coherent doc.  One thing I spotted is that in your internal clients example, the following:

mysvc.myns.myfed.svc.u7s.io

should be:

mysvc.myns.myfed.svc.cluster.local

You should also mention the positive case (where the service is found in the local cluster).

Thanks!  Lets move this into a MD doc.

Q

**Comment by madhusudancs** [comment]
I keep forgetting that I should @roberthbailey here, sorry about that. Please add any other stakeholders who should be involved in this discussion.

**Comment by madhusudancs** [comment]
@quinton-hoole I had a bunch of arguments about this second (alternative design), which is one of the reasons why I wanted to post it as a comment here and keep everything in one place. I am listing my arguments below. Let me know if you still like to move the second design and its arguments to a separate PR.

Here are my arguments about the second design.

* **Exposing internal only services externally?**

  - This design exposes all the federated services externally. What if users want to run cross-cluster services within their own network? Always programming an external Cloud DNS seems to be:
    - Unnecessary.
    - Dangerous (what are the security implications).
      - This makes services discoverable from outside, even when unintended.
      - Gives no control to the user to make the services not externally discoverable.
    - Against usability, esp. against, can I call it "tryability"?
      - Users can't try/explore the system within their own networks without setting up Cloud DNS.

* **Always requires users to register for a domain name**

  - Following up on the previous point, what if users want to keep the services internal to the federation and do not want to register a domain name for their federated services?

* **No way to ask explicitly ask for services in particular zones or regions.**

  - In this setup, if a user explictly asks for a servce in a particular zone they get back something else if the service does not exist in that zone.
  - This shouldn't be the case. Explicit requests should either return an answer if the service exists within the requested failure domain or say that the service doesn't exist.
  - In this case, we can't do a look-"up" in the tree, only a look-"down".

* **How to program SRV records?**

  - DNS spec for SRV records specifically says that the `target` of the service cannot be a CNAME.
  - How do we program SRV records with this design?

* **Cloud DNS limitations.**

  - There are quotas and limits and customers pay for each DNS zone.
  - We can't (or probably shouldn't) litter the Cloud DNS entries. People might want to use it for other purposes. One example that I came across recently was from a user who wanted to configure their PetSet shards in Cloud DNS for external clients. Just adding a ton of entries per service (upper bound of 216 records per service for a 100-cluster federation) makes it hard for clients.

* **Hard to explain it to users? How will they understand/debug?**

  - Following up on the previous point, it makes debugging harder for the users, esp. if they have additional DNS entries that they are managing within their projects.

* **The number of times the client resolver has to lookup for CNAME to follow the chain.**

  - Client resolver has to follow the CNAME chain, i.e. do several CNAME requests before it can connect to the service. The hope is that, caches work their magic.

* **Complexity in managing this tree for each and every federated service.**

  - There is quite a bit of complexity in managing this tree.
    - Every time a service shard goes up or down, we need to ensure that the tree is traversed and the entries are either inserted or purged into the tree and the
      tree remains consistent.

cc @quinton-hoole @thockin @ArtfulCoder @davidopp @nikhiljindal @brendandburns @roberthbailey

**Comment by quinton-hoole** [comment]
@madhusudancs And I have agreed to implement the [Alternative Design](https://github.com/kubernetes/kubernetes/pull/25107#issuecomment-218026648) for v1.3, and work on solutions to the points raised immediately thereafter.

**Comment by quinton-hoole** [comment]
@madhusudancs On further reflection, I have an idea that, if it works, at least partially addresses your first (and I believe most important concern) about exposing internal-only services externally (on a load balancer IP). 

cc @mfanjie @nikhiljindal @bprashanth @thockin FYI (comments welcome)

Services can have not only an external loadbalancer IP (a.k.a LoadBalancer Ingress IP), but also/alternatively an "ExternalIP", which, when configured, is assumed to be any IP address (or addresses) that are routed to any node in the cluster in question (through which kube-proxy based load balancing to the pods in the service is then performed).  http://kubernetes.io/docs/user-guide/services/#external-ips

So using GCE as an example, if the user configures such an IP inside their cluster, and it is reachable by nodes and pods in their GCE project (possibly in other clusters), then we can use that as the address of that service in that cluster, and it will only be reachable by other clusters in that GCE project/network, globally.

The challenge is how the user informs us of that IP in each cluster.  Initially I thought that we could just pick any or all of the node IP's in the cluster (which would be pretty easy), but that won't work - the traffic will go to the node, not the service, I would assume.  But, at least initially, we could allow the user to specify those per-cluster IP's via annotations on the federated service.  In future we could move this into real spec fields in the federated service, and could also auto-provision those external IPs.

**Comment by quinton-hoole** [comment]
PS:  This approach also works for NodePort services, and that gets around the problem of needing to tell the Ubernetes API of a different endpoint per cluster, as the same nodeport could be used in all clusters.
I chatted this through with @bprashanth and we think this will work.

**Comment by thockin** [comment]
Is this still alive?

**Comment by madhusudancs** [comment]
@thockin I don't know what the answer is to your question. Instead, I want your suggestion. A super simplified version of this proposal (in the original doc) is already implemented. The implementation as it exists today is a combination of a highly stripped down version of the doc + what is described in [one of the comments above](https://github.com/kubernetes/kubernetes/pull/25107#issuecomment-218026648). What should we do about this proposal?

**Comment by fejta** [comment]
I recommend you /close it unless it is providing value to you by staying open?

**Comment by johnbelamaric** [comment]
@chrisohaver and I have put together an alternative proposal that eliminates the need for an external DNS provider altogether. We would appreciate any comments, criticisms, and suggestions:

https://docs.google.com/document/d/1tvDE7oTAOKEPQr2HQDgCZQFDoVCoiy_ICQmvUIy6h5Y/edit?usp=sharing

**Comment by cblecker** [comment]
Design docs now live here: https://git.k8s.io/community/contributors/design-proposals
As this will never be merged into k/k, I'm going to close this. If this is still alive in some aspect, a new PR should be opened against the correct repo.

/remove-lifecycle frozen
/close

**Pull request closed without merging**